### PR TITLE
Alerting: Escape namespace and group path parameters

### LINKED ISSUE #76628: "Could not view rule" for alert rules provisioned via Grafana-Agent flow mode
### What happened?

I have Grafana Agent configured like so to load in any PrometheusRules added via GitOps:
```
      mimir.rules.kubernetes "local" {
        address = "http://infra-mimir-ruler:8080"
        tenant_id = "infrastructure"

        rule_namespace_selector {
          match_labels = {
            "kubernetes.io/metadata.name" = "grafana-ops",
          }
        }
        rule_selector {
          match_labels = {
            "grafana_agent" = "true",
          }
        }
      }
```
I can see the rules are loaded, visible, and evaluated in Grafana. When I attempt to dig deeper in to a specific alert, I am hit with an error and 404.

I get these errors when attempting to click on the "View alert rule" in the "Alerts" panel in a dashboard, and also when I click on the view button (the eye) in the Alert rules screen.
```
grafana logger=context userId=2 orgId=1 uname=some-user t=2023-10-16T13:14:42.220235142Z level=error msg="404 page not found\n" error="404 page not found\n" remote_addr=some-ip traceID=

grafana logger=context userId=2 orgId=1 uname=some-user t=2023-10-16T13:14:42.2219728Z level=info msg="Request Completed" method=GET path=/api/ruler/GazPBQB4z/api/v1/rules/agent/grafana-ops/coredns-forward/6cfef514-5766-4de4-9a36-407de1f1916c/coredns_forward status=404 remote_addr=some-ip time_ms=5 duration=5.417232ms size=47 referer="https://some-domain/alerting/Mimir%20(Infrastructure)/cri%24Mimir%20(Infrastructure)%24agent%1Fgrafana-ops%1Fcoredns-forward%1F6cfef514-5766-4de4-9a36-407de1f1916c%24coredns_forward%24CoreDNSForwardLatencyHigh%241378505054/view?returnTo=%2Falerting%2Flist" handler=/api/ruler/:DatasourceUID/api/v1/rules/:Namespace/:Groupname
```

### What did you expect to happen?

That I can see the details of that particular alert rule such as queries, labels, etc.

### Did this work before?

First time setting this up.

### How do we reproduce it?

1. Deploy Grafana, Mimir, and Grafana Agent (Configure it to to load PrometheusRules in to Mimir) in a Kubernetes cluster.
2. Setup a data source to Mimir where managed alerts are enabled.
3. Deploy a PrometheusRule in the cluster to be discovered by Grafana-Agent
4. The rule should then be discoverable, but viewing the details of the alert rule will cause a 404.

### Is the bug inside a dashboard panel?

Unlikely

### Environment (with versions)?

Grafana: v10.1.5
OS: Windows
Browser: Edge Chromium

### Grafana platform?

Kubernetes

### Datasource(s)?

Prometheus

### PR DESCRIPTION
**Pull request by SvenKirschbaum**
**What is this feature?**

Namespaces and groups of rules in Mimir can contain slashes in their name, which are currently not escaped when loading their configuration, which results in their detailed view being inaccessible in Grafana, as the API request returns 404. This PR ensures that path Parameters are always correctly escaped.

**Which issue(s) does this PR fix?**:

Fixes #76628

**Special notes for your reviewer:**

I am quite unsure whether I have chosen the best way to create the test cases, since I had to do some minor refactoring to make the functions testable in isolation. Would appreciate feedback on this.

### CODE REVIEW

**Review by JohnnyQQQQ — CHANGES_REQUESTED** [review]
Thanks for the contribution.

You can simplify the tests a bit. Right now you are interfacing the whole proxy and also adding private functions to a public interface.

You can just interface the function that you need for testing e.g. 

```
type requester interface {
	withReq(ctx *contextmodel.ReqContext, method string, u *url.URL, body io.Reader, extractor func(*response.NormalResponse) (any, error), headers map[string]string) response.Response
}
```

Than by default just use the normal proxy to do it

```
type requester interface {
	withReq(ctx *contextmodel.ReqContext, method string, u *url.URL, body io.Reader, extractor func(*response.NormalResponse) (any, error), headers map[string]string) response.Response
}

type LotexRuler struct {
	log log.Logger
	*AlertingProxy
	requester requester
}

func NewLotexRuler(proxy *AlertingProxy, log log.Logger) *LotexRuler {
	return &LotexRuler{
		log:           log,
		AlertingProxy: proxy,
		requester: proxy,
	}
}
```

When testing, you can manually build it and inject a function that asserts your code that is under test.

**Comment by SvenKirschbaum** [comment]
Thanks for taking a look @JohnnyQQQQ

I adapted the implementation of the tests accordingly.

File: pkg/services/ngalert/api/lotex_ruler.go [hunk 1]
```diff
@@ -43,15 +44,21 @@ var subtypeToPrefix = map[string]string{
 	Mimir:      mimirPrefix,
 }
 
+type requester interface {
```
Comment (JohnnyQQQQ) [inline pkg/services/ngalert/api/lotex_ruler.go:47]:
```suggestion
// The requester is primarily used for testing purposes, allowing us to inject a different implementation of withReq.
type requester interface {
```

**Review by JohnnyQQQQ — APPROVED** [review]
LGTM, thanks for taking care of this, great first contribution 👍

**Comment by JohnnyQQQQ** [comment]
@SvenKirschbaum if you rebase the build should pass and we can merge.

**Comment by SvenKirschbaum** [comment]
> @SvenKirschbaum if you rebase the build should pass and we can merge.

Done

**Pull request merged**